mv: update the backlogs when view updates finish

wmitros opened the pull request:
Currently, the backlog used for MV flow control is only updated
after we generate view updates as a result of a write request.
However, when the resources are no longer used, we should also
notice that to prevent excessive slowdowns caused by the MV
flow control calulating the delays based of an outdated, large
backlog.
This patch makes it so the backlogs are updated every time
a view update finishes, and not only when the updates start.
    
Fixes #18783

Similarly to https://github.com/scylladb/scylladb/pull/18646, without admission control (https://github.com/scylladb/scylladb/pull/18334), this patch doesn't affect much, so I'm marking it as backport/none

Tests: manual. Currently this patch only affects the length of MV flow control delay, which is not reliable to base a test on. A proper test will be added when MV admission control is added, so we'll be able to base the test on rejected requests

wmitros commented:
This patch only concerns the last commit. Is based on https://github.com/scylladb/scylladb/pull/18663 to avoid later conflicts

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9052/)
#### Build Details:
- Duration: 4 hr 5 min
- Builder: i-03a16ae8b4ef6afc6 (r5ad.8xlarge)

piodul commented:
> This patch only concerns the last commit. Is based on #18663 to avoid later conflicts

Is this necessary? I looked at the total diff of all 6 commits and the diff of the last commit and, at least to me, it looks like there shouldn't be conflicts.

Moreover, unless you are 100% sure that some change must go first I don't think it's a good idea to proactively rebase like this because the other patch blocks merging this one.

piodul commented on db/view/view.cc:
```diff
@@ -1793,6 +1793,8 @@ future<> view_update_generator::mutate_MV(
                 tracing::trace(tr_state, "Successfully applied local view update for {}", my_address);
                 return make_ready_future<>();
+            }).finally([this] {
+                _proxy.local().update_view_update_backlog();
```
Why not do it in the `then_wrapped` above? With the additional `finally` you are forced to do an additional allocation.

wmitros replied:
The update needs to be executed after the `sem_units` are destroyed which is a bit problematic because they're shared in both local and remote update path. I found we can do it a bit more verbosely without the additional allocation in the rebase.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9062/)
#### Build Failure:
```
[2024-05-22T08:27:29.866Z] FAILED: build/dev/db/view/node_view_update_backlog.hh.o 
[2024-05-22T08:27:29.867Z] clang++ -MD -MT build/dev/db/view/node_view_update_backlog.hh.o -MF build/dev/db/view/node_view_update_backlog.hh.o.d -std=c++20 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -U_FORTIFY_SOURCE -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_NO_CXX98_FUNCTION_BASE -DFMT_SHARED -I/usr/include/p11-kit-1 -std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-psabi -Wno-enum-constexpr-conversion -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -DSCYLLA_ENABLE_PREEMPTION_SOURCE -O2 -DSCYLLA_BUILD_MODE=dev -iquote. -iquote build/dev/gen  --include db/view/node_view_update_backlog.hh -c -o build/dev/db/view/node_view_update_backlog.hh.o build/dev/gen/empty.cc
[2024-05-22T08:27:29.867Z] In file included from <built-in>:1:
[2024-05-22T08:27:29.867Z] ./db/view/node_view_update_backlog.hh:55:5: error: no template named future; did you mean seastar::future?
[2024-05-22T08:27:29.867Z]     future<std::optional<update_backlog>> fetch_if_changed(std::optional<db::view::update_backlog> last_backlog);
[2024-05-22T08:27:29.867Z]     ^~~~~~
[2024-05-22T08:27:29.867Z]     seastar::future
[2024-05-22T08:27:29.867Z] /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/core/future.hh:1242:21: note: seastar::future declared here
[2024-05-22T08:27:29.867Z] class [[nodiscard]] future : private internal::future_base {
[2024-05-22T08:27:29.868Z]                     ^
[2024-05-22T08:27:29.868Z] 1 error generated.
```
#### Build Details:
- Duration: 10 hr
- Builder: i-0147a54f6104a64ed (m5ad.12xlarge)

mykaul commented:
@wmitros - please handle build failures.

wmitros commented:
I removed the commits that are not part of this patch - this patch should be easier to merge so we can fix the conflicts on the PRs that take longer to merge anyways

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9298/)
#### Build Details:
- Duration: 51 min
- Builder: i-020d7c845f58ed11e (m5d.8xlarge)

wmitros commented:
I simplified the code a bit and remove some duplication in the rebase. After reconsidering, if we're resetting the semaphore units on our own, we don't need to defer the backlog update to the end of the `then_wrapped` continuations, just after the `mutate_mv_locally`/`apply_to_remote_endpoints`. Because of that, I moved it to the start, before the future fail/success split

nyh commented on db/view/view.cc:
```diff
@@ -1777,8 +1777,10 @@ future<> view_update_generator::mutate_MV(
-                            sem_units] (future<>&& f) {
+                            sem_units, this] (future<>&& f) mutable {
                 --stats.writes;
+                sem_units = nullptr;
```
why is zeroing sem_units a few lines earlier important? Is it necessary for this patch, or just an improvement you thought of while doing this?

wmitros replied:
I want it to be zeroed before we update the view update backlog, otherwise we'll always be updating the backlog with a  value that's just about to change

nyh commented on db/view/view.cc:
```diff
@@ -1777,8 +1777,10 @@ future<> view_update_generator::mutate_MV(
+                            sem_units, this] (future<>&& f) mutable {
                 --stats.writes;
+                sem_units = nullptr;
+                _proxy.local().get_view_update_backlog();
```
I was a bit shocked to realize that this function, storage_proxy::get_view_update_backlog(), not only has the name "get", not only is "const", and not only has the same name as database::get_view_update_backlog - it actually **sets** the global backlog, doesn't just get the local one. This is so confusing :-( But of course it's not your fault. Maybe if we don't rename this function you should at least add a comment before it, something like `// update the node's view backlog from this shard's updated backlog.`

It's not pretty that we use atomic variables :-( But I looked, and it seems none of these operations are locked, so maybe it's fine performance-wise. I wonder if we *really* need to call it on every single completion. Maybe we could do it only if this node's backlog changed "a lot" since last time we updated the backlog or something... But maybe we just do this right now, and maybe reconsider later.

wmitros replied:
> I was a bit shocked to realize that this function, storage_proxy::get_view_update_backlog(), not only has the name "get", not only is "const", and not only has the same name as database::get_view_update_backlog - it actually **sets** the global backlog, doesn't just get the local one. This is so confusing :-( But of course it's not your fault. Maybe if we don't rename this function you should at least add a comment before it, something like `// update the node's view backlog from this shard's updated backlog.`

I'm planning to improve the naming problem anyway - I already have 8045bb5663f9b6ecabbf87140cbb0d4fee30153a in #18663 - I could move that commit here but maybe we can proceed with this current PR as it's quite simple and move on to #18663 

> It's not pretty that we use atomic variables :-( But I looked, and it seems none of these operations are locked, so maybe it's fine performance-wise. I wonder if we _really_ need to call it on every single completion. Maybe we could do it only if this node's backlog changed "a lot" since last time we updated the backlog or something... But maybe we just do this right now, and maybe reconsider later.

For now I'd like the backlog to be more reliable for https://github.com/scylladb/scylladb/pull/18334 which uses it. If we encounter some performance degradation, but we should for sure consider the approach you're mentioning

nyh commented on db/view/view.cc:
```diff
@@ -1777,8 +1777,10 @@ future<> view_update_generator::mutate_MV(
             local_view_update = _proxy.local().mutate_mv_locally(mut.s, *mut_ptr, tr_state, db::commitlog::force_sync::no).then_wrapped(
                     [s = mut.s, &stats, &cf_stats, tr_state, base_token, view_token, my_address, mut_ptr = std::move(mut_ptr),
-                            sem_units] (future<>&& f) {
+                            sem_units, this] (future<>&& f) mutable {
```
Please correct me if I'm wrong, but isn't this code a **local** mutation, i.e., in this current node, which is always synchronous and doesn't increase the view update backlog? So why do you need to touch the backlog here?

wmitros replied:
This doesn't increase the backlog only from the perspective of the client sending this request - the backlog does actually increase at the start of the process but it will always come back down before the request finishes, but in the meantime, the backlog will be higher for all other cilents.

If you were referring to the background tasks that only remote updates are generating, it's true that this branch does not create them, but it still holds on to some resources (in particular, to the mutation)

nyh commented on db/view/view.cc:
```diff
@@ -1813,7 +1815,9 @@ future<> view_update_generator::mutate_MV(
-                 sem_units, apply_update_synchronously] (future<>&& f) mutable {
+                 sem_units, apply_update_synchronously, this] (future<>&& f) mutable {
+                sem_units = nullptr;
+                _proxy.local().get_view_update_backlog();
```
Ditto on the question of why touch sem_units, and add a comment about the surprising meaning of get_view_update_backlog.

nyh reviewed (approved):
I'm approving because I think indeed the new code is more correct. However, I'm a bit at a loss in what kind of scenario the old un-perfect code produced wrong results. Is it just in the case where you had a lot of concurrent requests and then - suddenly - radio silence? Is it something you saw in a small test and wanted to fix, or something you believe really affected real workloads?

The cost of this improvement is extra unlocked (good) relaxed-order (good) atomic operations, but I think those are cheap in x86 (I didn't check about the cache line alignment of these counters..), so let's let it slides. We anyway had these operations when starting the view update, so this change at most doubles them, not more.
